Apply workaround to mask issues with repainting some controls

a-wi opened the pull request:
This is related to issue #22629.

vadz commented:
@a-wi should we merge this or do you still plan to make any changes here?

a-wi commented:
I'm done with this workaround.
Any feedback from those who could reproduce the issue #22629 (@whalley, @codesquid) would be welcome.

whalley commented:
Apologies for delay in testing. I'm away at present so will have to look at this after the 12th Dec.

a-wi commented:
@whalley
Since restoring graphics context attributes didn't help so I think next suspect are font settings. It looks like previous (emulated) version of ResetClip() masked some font issues and new (real) version just revealed them.
I've added some more diagnostic logging to see what are font settings used in wxTreeCtrl and it would be helpful if you could pull new code from this PR, run it in debug mode and check what is logged when you are reproducing the issue.

vadz commented:
Could we apply the known to work workaround (there was one, right?) to master before continuing to look for a better one? I think we do need to do at least something to avoid regressions in 3.3.0, don't we?

whalley commented:
Apologies, not had time to spend much more testing on this due to other commitments.... Reverting according to this https://github.com/wxWidgets/wxWidgets/issues/22629#issuecomment-1315175124 works for me.

a-wi commented:
I we are not in a hurry I wouldn't apply crippled workaround to master. I am going to do more investigations with this issue.

vadz commented:
> I we are not in a hurry I wouldn't apply crippled workaround to master.

We're not in a hurry (this would need to be fixed before 3.3.0, but I don't know when it's going to happen yet), but I don't like having master broken for longer than necessary, so I'd prefer to apply a workaround which avoids the regression already. Of course, absolutely nothing prevents us from reverting it and applying a better one as soon as it becomes available, so I don't see what we lose here. OTOH having this problem in master for a couple of months already is not ideal IMO.

a-wi commented:
I'm also not comfortable with this issue being open for so long but the catch with this issue is that it's not universally reproducible and it's hard to isolate the root cause. And you know how looks debugging/fixing the issue if you can't reproduce it.
The catch with so called workaround is that this s not a workaround per se but rather a mask hiding the real problem. And it's not a neutral mask because it has side effects.  And what is worse we have no idea where the problem lies so this mask would be applied completely blindly. So I maybe having an open issue is not that bad option. At least there is a motivation to do something.
I will try to dig more in a new environment where maybe I would be able to reproduce the issue, for a good start.

a-wi committed:
Fix resetting clipping region in wxMacCoreGraphicsContext

Instead of resetting entirely the clipping region of CGContext
we should reset the clipping region to the one that was
applied initially to CGContext when wxMacCoreGraphicsContext
was initialized.

a-wi committed:
xApply workaround to mask issues with repainting wxTreeCtrl on wxOSX

Emulating the reset of the clipping area, used originally prior to
10.13, apparently has some side-effects which help in masking
the issue with repainting wxTreeCtrl.
The root cause of the issue is not yet known, but resetting
the clipping region before drawing the items seems to help.

See #22629.

vadz commented on src/generic/treectlg.cpp:
```diff
@@ -2576,6 +2576,10 @@ void wxGenericTreeCtrl::PaintItem(wxGenericTreeItem *item, wxDC& dc)
 
     int offset = HasFlag(wxTR_ROW_LINES) ? 1 : 0;
 
+#ifdef __WXOSX__
+    dc.DestroyClippingRegion();
+#endif // __WXOSX__
```
This is really weird, where is this clipping region set? We only use `wxDCClipper` in this file itself, so how can we have any clipping region here?

a-wi replied:
The issue itself is weird - how is it possible that call to native CGContextResetClip(() impacts font weight  - so the workaround is weird too.
But this is still experiment, so maybe in the next turn something less weird will be found.

a-wi commented:
Won't be implemented.
